Add express comparison docs (#1349)
encoredev/encore: docs/how-to/express-comparison.md

---

> seotitle: Migrating your Express backend to Encore.ts

Reviewer: ```suggestion
seotitle: Migrate from Express to Encore.ts
```

seodesc: Learn how migrate your Express app over to use Encore.ts for better performance and improved development tools.
title: Migrating from Express to Encore.ts
---

> If you have an existing [Express](https://expressjs.com/) application and want to migrate it to Encore.ts, this guide is
> for you. This guide can also serves as a comparison between the two frameworks.

Reviewer: ```suggestion
If you have an existing app using [Express](https://expressjs.com/) and want to migrate it to Encore.ts, this guide is
for you. This guide can also serve as a comparison between the two frameworks.
```

## Why migrate to Encore.ts?

> Express is a great choice for building simple APIs, but as your application grows, you will eventually run into
> limitations. Because of the big community around Express, there are a lot of plugins and middleware available, but this
> can also make it hard to find the right tools for your use case. This also means that you will need to maintain a
> lot of dependencies.

Reviewer: ```suggestion
Express is a great choice for building simple APIs, but as your application grows you will likely run into
limitations. There is a large community around Express, which is great as there are a lot of plugins and middleware available, but relying heavily on plugins can make it hard to find the right tools for your use case. It also means that you will need to maintain a lot of dependencies.
```

> Encore.ts is a web framework that aims to make it easier to build robust and type-safe backends with
> TypeScript. Encore.ts has 0 npm dependencies, is built with performance in mind and has a lot of built-in features for
> building production ready backends. You can deploy an Encore.ts app to any hosting service that accepts Docker
> containers, or use [Encore Cloud Platform](/use-cases/devops-automation) to fully automate your DevOps and
> infrastructure.

Reviewer: ```suggestion
Encore.ts is a framework that aims to make it easier to build robust and type-safe backends with
TypeScript. Encore.ts has 0 npm dependencies, is built with performance in mind, and has a lot of built-in features for building production ready backends. You can deploy an Encore.ts app to any hosting service that accepts Docker containers, or use [Encore Cloud Platform](/use-cases/devops-automation) to fully automate your DevOps and infrastructure.
```

### Performance

> Unlike a lot of other Node.js frameworks, Encore.ts is not build a top of Express. Instead, Encore.ts has a
> high-performance runtime, with a multi-threaded, asynchronous event loop written in Rust. The Encore Runtime handles all
> I/O like accepting and processing incoming HTTP requests. This runs as a completely independent event loop that utilizes
> as many threads as the underlying hardware supports. The result of this is that Encore.ts performs **9x faster** than
> Express. Learn more about the [Encore.ts Runtime](/blog/event-loops).

Reviewer: ```suggestion
Unlike a lot of other Node.js frameworks, Encore.ts is not built on top of Express. Instead, Encore.ts has its own
high-performance runtime, with a multi-threaded, asynchronous event loop written in Rust. The Encore Runtime handles all I/O like accepting and processing incoming HTTP requests. This runs as a completely independent event loop that utilizes as many threads as the underlying hardware supports. The result of this is that Encore.ts performs **9x faster** than Express. Learn more about the [Encore.ts Runtime](/blog/event-loops).
```

### Built-in benefits

When using Encore.ts you get a lot of built-in features:

- [Pub/Sub integrations](/docs/ts/primitives/pubsub)
- [Type-safe API schemas](/docs/ts/primitives/services-and-apis)
- [API Client generation](/docs/develop/client-generation)
- [Secrets management](/docs/ts/primitives/secrets)
- [CORS handling](/docs/ts/develop/cors)
- [Infrastructure integrations](/docs/deploy/infra)
- [Database integrations](/docs/ts/primitives/databases)
- [Architecture Diagrams](/docs/observability/encore-flow)
- [Service Catalog](/docs/develop/api-docs)
- [Request validation](/blog/event-loops)
- [Cron Jobs](/docs/ts/primitives/cron-jobs)
- [Local tracing](/docs/observability/tracing)
- [Local Development Dashboard](/docs/observability/dev-dash)

## App architecture

> Encore.ts is like Express in that it is unopinionated in how you structure your code.
> 
> We recommend however you use one Encore app for your entire backend application. This lets Encore
> build an application model that spans your entire app, necessary to get the most value out of many features like
> [distributed tracing](https://encore.dev/docs/observability/tracing)
> and [Encore Flow](https://encore.dev/docs/observability/encore-flow).
> 
> Learn more in our [App Structure docs](https://encore.dev/docs/ts/develop/app-structure).

Reviewer: ```suggestion
Like Express, Encore.ts in not opinionated in how you structure your code.

It is recommended to use one Encore app for your entire backend application, as this lets Encore
build an application model that spans your entire system. This is necessary to get the most value out of many features like [distributed tracing](https://encore.dev/docs/observability/tracing) and [Encore Flow](https://encore.dev/docs/observability/encore-flow).

Learn more in our [App Structure docs](https://encore.dev/docs/ts/develop/app-structure).
```
